### PR TITLE
Add new combo and combo skip to `HitObject`

### LINKED ISSUE #112: New Combos not parsed
Hit objects don't parse new combo from the type. I need new combos for my project.

New combos are stored in bit index 2 of the type:
https://osu.ppy.sh/wiki/en/Client/File_formats/osu_%28file_format%29#type

### PR DESCRIPTION
**Pull request by OliBomby**
Seemed like this was missing. I need new combos for my project.

Spinners dont parse the combo skip value because this mimics osu! stable behaviour.
Hold notes also dont parse the combo information.

Closes #112

### CODE REVIEW

File: slider/tests/test_beatmap.py [hunk 1]
```diff
@@ -142,6 +142,8 @@ def test_parse_section_hit_objects(beatmap):
 def test_hit_objects_stacking():
```
Comment (Stedoss) [inline slider/tests/test_beatmap.py:142]: Might be worth asserting these in the `test_parse_section_hit_objects` function above too -
```py
assert hit_objects_0.new_combo
assert not hit_objects_0.combo_skip
```

Comment (OliBomby) [reply slider/tests/test_beatmap.py:144]: True, added

**Comment by tybug** [comment]
I've made some changes here. Before, this pull included breaking changes: `from slider import Circle; c = Circle(...)` is part of the public api and any new parameters should have a default value (unless we want to bump the major version, which I don't particularly want to right now).

I've also gone ahead and parsed the combo skip for every hitobject. Even if combo skip values are unused for certain hitobjs by osu stable, as long as they are present in the `type` information of the beatmap I think it is more correct for us to parse them according to spec and provide them. A consumer can then decide how to use the combo skip number in the context of different hitobjs.

I haven't tested this besides making sure code tests pass, so probably worth a double check to ensure I didn't break anything.

File: slider/beatmap.py [hunk 2]
```diff
@@ -297,6 +306,21 @@ def _time_modify(self, coefficient):
...
         return type(self)(**kwargs)
 
+    def _get_combo_bits(self):
+        # bit numbers below are zero indexed.
+
+        # type code (bits number 0, 1, 3, and 7)
+        bits = self.type_code
+        # new combo (bit 2)
+        bits |= self.new_combo << 2
+        # first (MSB) of combo_skip (bit 4)
```
Comment (OliBomby) [inline slider/beatmap.py:316]: Why did you make the code like this. Its so hard to understand. Also the comment is wrong

Comment (tybug) [reply slider/beatmap.py:316]: I'm not particularly happy with it either, do you have a better suggestion? And can you elaborate on how the comment is wrong?

Comment (OliBomby) [reply slider/beatmap.py:316]: it says second bit of combo skip twice.
Please just use my implementation of this method. It was good as it was.

File: slider/beatmap.py [hunk 3]
```diff
@@ -420,7 +444,25 @@ def parse(cls, data, timing_points, slider_multiplier, slider_tick_rate):
         else:
             raise ValueError(f'unknown type code {type_!r}')
 
-        return parse(Position(x, y), time, hitsound, rest)
+        # new combo info is in second bit (0-indexed)
+        new_combo = bool(type_ & (1 << 2))
+        # abusing the int constructor which takes a string and a base. blame
+        # python for not having sane bitarray handling.
+        bitarray = []
```
Comment (OliBomby) [inline slider/beatmap.py:451]: I'm a bit shocked by this implementation of getting the combo skip. My implementation already calculated it in one line...

Comment (OliBomby) [reply slider/beatmap.py:451]: if you know that integers are basically bit arrays, you can just do `type_ & 112` to select bits 4, 5, and 6, and then do `>> 4` to shift the 3 bits to the right, giving you the combo skip value.

File: slider/beatmap.py (hunk 2 above)
Comment (tybug) [reply slider/beatmap.py:316]: on second look this does seem to be more confusing than it's worth (too explicit for its own good). I've changed to your implementation

File: slider/beatmap.py (hunk 3 above)
Comment (tybug) [reply slider/beatmap.py:447]: the bitarray stuff is just stupid, don't get me wrong. would be very happy to get rid of that. I was mostly trying to make the code more explicit - it's pretty hard to tell where the magic number `112` comes from in `type_ & 112`.

I changed it to your version, but with an explicit mask - does that read better?

I don't often work with bit operations, so things which are likely idiomatic and obvious (shifting to get the desired portion of the integer) seem confusing to me at first and worth replacing (even if they aren't). I agree with your code more now that I've had a chance to brush up on my bitwise stuff.

File: slider/beatmap.py [hunk 4]
```diff
@@ -257,12 +257,22 @@ class HitObject:
         Unknown currently.
     """
     time_related_attributes = frozenset({'time'})
-
-    def __init__(self, position, time, hitsound, addition='0:0:0:0:'):
+    type_code = 0
```
Comment (tybug) [inline slider/beatmap.py]: hmm...it's already hard for consumers to subclass `HitObject` in a useful way as class references are hardcoded in `HitObject#parse`, so I'm not sure how much adding this attribute this matters for consumers when it's already set for all internal members. I guess it's not hurting anyone, though.

I have however changed it to `None`, as `type_code = 0` is likely to silently succeed in places an int is expected and break things if it ever falls through to a subclass.

Comment (OliBomby) [reply slider/beatmap.py]: I specifically made it an integer so `_get_type_bits()` wouldn't throw a `TypeError` when its trying to use the `|` operator on a NoneType and int

Comment (tybug) [reply slider/beatmap.py]: well, my point is that method throwing a type error when `type_code` isn't set is good. That's what I meant by things "silently breaking" - if we ever subclass `HitObject` and forget to set a `type_code`, that should show up as a failure, not silently "work" (but produce incorrect results).

Comment (OliBomby) [reply slider/beatmap.py]: yes i realized that now

**Comment by tybug** [comment]
thanks for the `HoldNote` fix 👍 

this looks good to me

**Review by tybug — APPROVED** [review]

**Pull request merged**